### PR TITLE
Joystick Hotplugging

### PR DESCRIPTION
**Pull request by jaxxzer**
I have tested this extensively under Windows and Ubuntu with multiple joysticks being connected/disconnected. When a Joystick is unplugged, the manual control messages stop, which fixes https://github.com/mavlink/qgroundcontrol/issues/3980.

An **important note** here, when a joystick is removed, the JoystickSDL thread associated with it continues to run, and the pointer to the object is lost. If the joystick is plugged back in, a new thread is started. I am not sure how to delete the thread safely, perhaps someone can lend a hand [here](https://github.com/mavlink/qgroundcontrol/pull/4477/files#diff-772bdcd04c41931b045c56907653a23cR92).

I have also had some infrequent issues with segfaults on SDL_JoystickClose, it seems like the pointer returned by SDL_JoystickOpen is not always valid. This is even the case when calling SDL_JoystickOpen, then immediately calling SDL_JoystickClose. It appears to me to be an issue with SDL. This has only popped up as an issue outside of normal use cases with multiple joysticks plugged in, where joysticks were being opened and closed very many times.

### CODE REVIEW

**Comment by jaxxzer** [comment]
That appears to do the trick. Thank you for the hint, I will update this after I have worked it in.

File: src/Joystick/JoystickManager.cc [hunk 1]
```diff
@@ -92,11 +92,10 @@ void JoystickManager::_setActiveJoystickFromSettings(void)
             qCDebug(JoystickManagerLog) << "Releasing joystick:" << i.key();
             if (_activeJoystick && !newMap.contains(_activeJoystick->name())) {
                 qCDebug(JoystickManagerLog) << "\twas active";
-                _activeJoystick->stopPolling();
-                _activeJoystick = NULL;
+                setActiveJoystick(NULL);
             }
-            // FIXME: Don't leave this hanging! We need to free the object.
-            //i.value()->deleteLater();
+            i.value()->wait(1000);
```
Comment (NaterGator) [inline src/Joystick/JoystickManager.cc]: I believe you also want `i.value()->stopPolling();` here before the wait...

**Comment by DonLakeFlyer** [comment]
> } else {
>     i.value()->stopPolling();
> }
> i.value()->wait(1000);
> i.value()->deleteLater();

Isn't there some sort of signal you can wait on with QSignalSpy instead of doing a wait?

File: src/Joystick/JoystickManager.cc [hunk 2]
```diff
@@ -146,3 +187,31 @@ void JoystickManager::setActiveJoystickName(const QString& name)
...
     setActiveJoystick(_name2JoystickMap[name]);
 }
+
+void JoystickManager::_updateAvailableJoysticks(void)
+{
+#ifdef __sdljoystick__
+    SDL_Event event;
+    while (SDL_PollEvent(&event)) {
+    	switch(event.type) {
+        case SDL_QUIT:
```
Comment (DonLakeFlyer) [inline src/Joystick/JoystickManager.cc]: Coding style issues here. ```case``` should line up with ```switch```

File: src/Joystick/JoystickManager.cc [hunk 3]
```diff
@@ -146,3 +187,31 @@ void JoystickManager::setActiveJoystickName(const QString& name)
...
+            break;
+        case SDL_JOYDEVICEADDED:
+            qCDebug(JoystickManagerLog) << "Joystick added:" << event.jdevice.which;
+            _setActiveJoystickFromSettings();
+            break;
+        case SDL_JOYDEVICEREMOVED:
+            qCDebug(JoystickManagerLog) << "Joystick removed:" << event.jdevice.which;
+            _setActiveJoystickFromSettings();
+            break;
+		default:
```
Comment (DonLakeFlyer) [inline src/Joystick/JoystickManager.cc]: ```default``` should line up with ```switch```

**Comment by DonLakeFlyer** [comment]
Generally looks ok. But I don't have a joystick with me on the road to test out. If it's working fine for Nate as well then I'm ok with this.

**Comment by DonLakeFlyer** [comment]
@NaterGator Can you run/try this as well?

**Comment by NaterGator** [comment]
I'll try to give it a go tomorrow with some equipment but I'm actually doing some remote field work this week and not near my development machine at the office. Will report back tomorrow.

**Comment by DonLakeFlyer** [comment]
@NaterGator Have you had time for this yet?

**Comment by DonLakeFlyer** [comment]
@jaxxzer I'm back in town now. Can you fix conflict. Then I'll do a double-check test and merge.

**Comment by DonLakeFlyer** [comment]
LGTM

**Pull request merged**